lib/request-proxy-send.js

```diff
@@ -0,0 +1,279 @@
+        ' now meant for {newDestinations}.',
+    keys: null,
+    origDestination: null,
+    newDestinations: null
+});
+
+var RETRY_SCHEDULE = [0, 1, 3.5];
+
+function RequestProxySend(opts) {
+    this.ringpop = opts.ringpop;
+    this.requestProxy = opts.requestProxy;
+    this.keys = opts.keys;
+    this.channelOpts = opts.channelOpts;
+    this.request = opts.request;
+    this.retries = opts.retries;
+
+    this.retrySchedule = this.retries.schedule || RETRY_SCHEDULE;
+    this.maxRetries = numOrDefault(this.retries.max, this.retrySchedule.length);
+    this.maxRetryTimeout = this.retrySchedule[this.retrySchedule.length - 1] * 1000;
+
+    this.destinations = [this.channelOpts.host];
+    this.errors = [];
+    this.numRetries = 0;
+    this.reqStartTime = Date.now();
+    this.retryStartTime = null;
+    this.timeout = null;
+}
+
+RequestProxySend.prototype.abortRetry = function abortRetry(dests, callback) {
+    this.ringpop.stat('increment', 'requestProxy.retry.aborted');
+    this.ringpop.logger.warn('ringpop request proxy retry aborted after keys destinations diverged', {
+        local: this.ringpop.whoami(),
+        destOrig: this.channelOpts.host,
+        destNew: dests,
+        destinations: this.destinations,
+        reqMethod: this.request.obj.method,
+        reqUrl: this.request.obj.url
+    });
+
+    this.ringpop.emit('requestProxy.retryAborted');
+
+    callback(RetryDivergenceError({
+        keys: this.keys,
+        origDestination: this.channelOpts.host,
+        newDestinations: dests
+    }));
+};
+
+RequestProxySend.prototype.attemptRetry = function attemptRetry(callback) {
+    this.numRetries++;
+
+    // TODO Support this case too
+    var dests = this.lookupKeys(this.keys);
+    if (dests.length > 1) {
+        this.abortRetry(dests, callback);
+        return;
+    }
+
+    this.ringpop.stat('increment', 'requestProxy.retry.attempted');
+    this.ringpop.emit('requestProxy.retryAttempted');
+
+    var newDest = dests[0];
+
+    // Nothing rebalanced.
+    if (newDest === this.channelOpts.host) {
+        this.send(this.channelOpts, callback);
+        return;
+    }
+
+    this.rerouteRetry(newDest, callback);
+};
+
+RequestProxySend.prototype.destroy = function destroy() {
+    clearTimeout(this.timeout);
+};
+
+RequestProxySend.prototype.getRawHead = function getRawHead() {
+    return rawHead(this.request.obj, {
+        checksum: this.ringpop.membership.checksum,
+        keys: this.keys
+    });
+};
+
+RequestProxySend.prototype.getStrHead = function getStrHead() {
+    return strHead(this.request.obj, {
+        checksum: this.ringpop.membership.checksum,
+        keys: this.keys
+    });
+};
+
+RequestProxySend.prototype.handleMaxRetriesExceeded = function handleMaxRetriesExceeded(callback) {
+    this.ringpop.stat('increment', 'requestProxy.retry.failed');
+    this.ringpop.logger.warn('ringpop request proxy retry exceeded max retries and failed', {
```

Comment:
A lot of these logger statements have a large object.

Maybe in the constructor we should pre-allocate a well defined `logContext` object with the majority of these properties.

This will make the logging context very clear in the constructor and make the actual log sites simpler.

It also allows us to inline a couple of these helper methods.